doc/source/reference/c-api/types-and-structures.rst

```diff
@@ -71,19 +71,17 @@ PyArray_Type and PyArrayObject
    member of this structure contains a pointer to the :c:data:`PyArray_Type`
    typeobject.
 
-.. c:type:: PyArrayObject
-            NPY_AO
-
-   The :c:type:`PyArrayObject` C-structure contains all of the required
-   information for an array. All instances of an ndarray (and its
-   subclasses) will have this structure.  For future compatibility,
-   these structure members should normally be accessed using the
-   provided macros. If you need a shorter name, then you can make use
-   of :c:type:`NPY_AO` (deprecated) which is defined to be equivalent to
-   :c:type:`PyArrayObject`. Direct access to the struct fields are
-   deprecated. Use the ``PyArray_*(arr)`` form instead.
-   As of NumPy 1.20, the size of this struct is not considered part of
-   the NumPy ABI (see note at the end of the member list).
+.. c:type:: PyArrayObject_fields
+
+   The :c:type:`PyArrayObject_fields` C-structure contains all of the
+   required information for an array. All instances of an ndarray (and
```

Comment:
I don't think this is correct. It holds references to "legacy" functionality. @seberg might have a better suggestion for language, he managed this API transition.

Reply:
@ngoldbaum I am not sure I understand your comment.

> he managed this API transition.

Nope, that is before my time even :).  Maybe `PyArrayObject_fields` should just say that it is an implementation detail for accessing the `PyArrayObject` struct when it is opaque (due to new API use).  That is, to implement macros and static-inline functions which are the preferred way of accessing array fields/information.

I would be happy to keep the actual struct definition under `PyArrayObject` and just note that this struct is opaque without deprecated API and direct access is discouraged since it may allow future API improvements.

(Frankly, I am not really enthusiastic about ever doing it, but maybe... in the end I doubt we need to break ABI here unless we break API in some way. And if we break API as well for some field, then not sure the drepcation will actually help much :))

Reply:
Thanks for your feedback! I've moved the definition back to `PyArrayObject` and submitted a new PR.